### ISSUE TITLE
Library seems incompatible with stock lz4

### ISSUE THREAD

**Issue by istathar** [mwotton/lz4hs#10]
I compressed a file using the `lz4` command line utility, then attempted to retrieve it using

``` haskell
main = do
    x' <- S.readFile "/home/andrew/Book.pdf.lz4"

    let y' = decompress x'

    putStrLn $ show $ S.length x'
    putStrLn $ show $ fmap S.length y'
```

to which I got

> 1562102
> Nothing

I'm assuming this has to do with the Haskell bindings deciding to write (and subsequently expect on reading) some extra bytes to do with the length of the encoded message. I realize there's a design problem here, but surely being incompatible with the real LZ4 that other people are sending us would be considered a drawback?

Is there a way to avoid this problem? I'm assuming the answer is going to be "use the FFI directly" but I figured I should check to find out if there was deliberate.

AfC

**Comment by mwotton** [comment]
@thoughtpolice - had a look at this, not quite sure what's going on with the framing on line 166 of LZ4.hsc. Could you have a look?

**Comment by Cyan4973** [comment]
Note that a file compression format (also suitable for streaming scenario) has been defined and documented.
http://fastcompression.blogspot.fr/2013/04/lz4-streaming-format-final.html
(it's also in the root directory on code.google.com repository).

All implementations are invited to adopt it to ensure compatibility between versions.

**Comment by istathar** [comment]
Sorry to be disruptive, but the Haskell library doing it's own thing is a pretty serious incompatibility. I mean, sure, it's inconvenient that it doesn't know its own length and that you have to transmit it externally, but that's the reality with almost everything: Protocol Buffers suffer from the same "feature",  but they do so because HTTP does Content-Length; ØMQ does message length; hell, I just wrote something which has a nice tightly encoded 2-byte prefix to be a frame for such things on disk [I can always library-ize that if you need something :)].

But given the streaming format upstream defined as @Cyan4973 points out, surely that would be a better way to solve the problem?

AfC

**Comment by mwotton** [comment]
@afcowie yes, agreed. I consider this a bug, but don't currently have time to implement @Cyan4973 's format. Patches gratefully accepted.

**Comment by cartazio** [comment]
i may take some time to explore this, but not guarantees. if i've not done it by start of april, holler :)

**Comment by Cyan4973** [comment]
Not sure if it's useful, but just in case : 

functions to compress files using the official streaming format are now directly accessible,
into lz4io.h / lz4io.c

It's supposed to help bindings. Feedbacks are welcomed.

**Comment by mpickering** [comment]
Is there any workaround to this issue? I need to decompress some standard lz4 but the decompression function exposed by the library expects the size to be written at the start of the bytestring.

**Comment by mwotton** [comment]
@mpickering sorry, still haven't had time. @cartazio how's your work queue?

**Comment by mpickering** [comment]
This isn't critical at all for me. In the end I had to use a foreign call to the javascript library as I was using ghcjs.

**Comment by cartazio** [comment]
I'll have a go at cleaning / modernizing stuff today or later this week if
@mwotton eg al review the change after I pr it

**Comment by mwotton** [comment]
Happy to review.
